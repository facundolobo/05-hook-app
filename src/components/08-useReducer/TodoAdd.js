import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    //estado del formulario
    const [{description}, handleInputChange, reset]=useForm({
        description: ''

    });


    const handleSubmit = (e) =>{

        e.preventDefault();//evista le posteo de formulario o reflesh

        //funcion si no escribio nada no lo guarda
        if(description.trim().length <=1){
            return;
        }


        const newTodo ={
            id: new Date().getTime(),
            desc: description, 
            done: false  //done = hecho
        };

        handleAddTodo( newTodo );
        reset(); //borra lo que estaba en el input

    }



    return (
        <>
            <h4>
                    agregar TODO
                </h4>
                <hr/>

                {/* formulario apra agregar una tarea */}
                <form onSubmit={handleSubmit}>
                    <input 
                        type="text"
                        name="description"
                        className="form-control"
                        placeholder="aprender..."
                        autoComplete="off"

                        value={description}
                        onChange={handleInputChange}
                    />
                    <button 
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                    >
                        Agregar
                    </button>
                </form>
        </>
    )
}
