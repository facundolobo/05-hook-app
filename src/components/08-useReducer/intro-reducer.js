const initialState = [{
    id:1,
    todo: 'Comprar pan',
    done: false //stado de la tare, no se realizo 

}]
const  todoReducer = ( state = initialState , action ) => { //enviamos estado y una acci√≥n

    if(action?.type === 'agregar'){

        return [...state, action.payLoad];
    }

    return state; 
}

let todos= todoReducer();

//funcion agregar un nuevo todo
const newTodo ={
    id:2,
    todo: 'Comprar leche',
    done: false 
}
//------------------------

//funcion para agregar a la funcion
const agregarTodoAction ={
    type:'agregar',
    payLoad: newTodo
}
//--------------------

todos = todoReducer (todos, agregarTodoAction);

console.log( todos );