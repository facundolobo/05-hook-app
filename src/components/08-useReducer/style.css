body{
    padding: 50px;
}
p{
    cursor: pointer;
    margin: 0;
}

.list-group-item{
    align-items:center;
    display : flex;
    justify-content:  space-between;
}

.complete{ /* si agrego esta clase a un <p> entocnes aparecera tachado*/
    text-decoration: line-through;
}
