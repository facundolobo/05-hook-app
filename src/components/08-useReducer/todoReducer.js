export const todoReducer = (state = [] , action)=>{

    switch (action.type) {
        case 'add':
            
            return  [...state,action.payload];

        case 'delete':
            return state.filter( todo => todo.id !== action.payload);  //compara con el id del payload enviado
 
        case 'toggle': //forma abreviada
            return state.map(todo => 
                (todo.id === action.payload)
                    ? {...todo,done:!todo.done}
                    :todo
            );

        case 'toggle-old':
            return state.map(todo => { //recorre todos los elementos como un for
                if(todo.id === action.payload){
                    return {...todo,done:!todo.done} //cambiamos el valore
                }else{
                    return todo; //devemos devolverlo igual sino regresara un undefine
                }

            })


            default:
            return state;
    }
}