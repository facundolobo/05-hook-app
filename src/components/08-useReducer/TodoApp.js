import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';
import {TodoItem} from './TodoItem.js'
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';

import './style.css';



const init =() =>{
  
    //si existe el archivo en local store lo regresa, sino []
    return JSON.parse(localStorage.getItem('todos')) || [];;
  
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React', 
    //     done: false  //done = hecho
    // }]
}


export const TodoApp = () => {

    
    const [todos, dispatch] = useReducer(todoReducer, [], init);

   
    useEffect(()=>{
        localStorage.setItem('todos', JSON.stringify(todos))
    },[todos]);

    const handleDelete = (todoId)=>{
        //e.preventDefault();//evista le posteo de formulario o reflesh
        console.log(todoId);

        //crear la action
        
        
        
        const action = {
            type:'delete',
            payload: todoId //envia el id de la tarea asociada
        }
        // dispatch
        dispatch( action );
    }   

    //console.log(description);
    const handleToggle=(todoId)=>{
        dispatch({ //forma corta 
            type: 'toggle',
            payload:todoId
        })
    }

    const handleAddTodo=(newTodo)=>{
        
        dispatch( {
            type:'add',
            payload: newTodo
        } );
    }



    return (
        <div className="row"> 
           <h1>TodoApp ({todos.length})</h1> 
           <hr/>
            
            <div className="col-7">
            
                <TodoList
                todos={todos} 
                handleDelete = {handleDelete} 
                handleToggle={handleToggle} 
                />
            
            </div>
            
            <div className="col-5">
                <TodoAdd
                handleAddTodo={handleAddTodo}
                />
            </div>

            
        </div>    
        
            

    )
}
