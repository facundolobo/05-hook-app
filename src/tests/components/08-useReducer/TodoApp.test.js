import React from 'react';
import { act } from "@testing-library/react";
import { shallow } from "enzyme";

import { mount } from "enzyme"; //debemos usar una libreria mas antigua para q funciones el mount

import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { demoTodos } from "../../fixtures/demoTodos";

describe('Pruebas en <TodoApp/>', () => {
    const wrapper= shallow(<TodoApp/>)
    test('debe de mostrarce correctamente ', () => {
        expect(wrapper).toMatchSnapshot();
    })
   
    // test('debe de agregar un TODO ', () => {
    //     const wrapper= mount(<TodoApp />);
        
        
        
        
    //     //cuando necesitamos probar la app en contexto, osea en general
    //     act(()=>{
    //         //nos vinculamos a handleAddTodo
    //         // de todoApp y enviamos un un objeto demoTodos[0]
    //         wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0])
    //         wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1])
    //     });
    //     expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)')
    //     expect(localStorage.setItem).toBeCalledTimes(2);
       
    // })
    test('Debe de borrar un todo ', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0])
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id)
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)')
    })
    
})
