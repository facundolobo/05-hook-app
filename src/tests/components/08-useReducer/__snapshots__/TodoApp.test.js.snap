// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp/> debe de mostrarce correctamente  1`] = `
<div
  className="row"
>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="col-7"
  >
    <TodoList
      handleDelete={[Function]}
      handleToggle={[Function]}
      todos={Array []}
    />
  </div>
  <div
    className="col-5"
  >
    <TodoAdd
      handleAddTodo={[Function]}
    />
  </div>
</div>
`;
